Reconstruct_sdms

TukisTukis opened the pull request:
Added a function to produce reconstructed points with SDMs from a set of angles and distances

Commit:
Moved angle functions to own module.

Commit:
Fixed new RLS_SDR function.

Commit:
Refactoring.

Commit:
Updated docs.

Commit:
Updated version.

Commit:
Updated doc release.

Commit:
Merge pull request #1 from TukisTukis/type/feature

new functions for sdms

Commit:
Moved abs_angles to base class.

Commit:
Merge branch 'master' of github.com:LCAV/pylocus

Commit:
Simplified functions, added new get_V.

duembgen commented:
This goes in the wrong direction. I made a new pull request from TukisTukis to LCAV and assigned you to it.

Commit:
Merge pull request #3 from TukisTukis/type/sdm_pipe

Created reconstruct_smds

Commit:
cleaned up shapes.

Commit:
Added execute_method wrapper.

Commit:
Minor changes.

Commit:
Merge branch 'master' of github.com:LCAV/pylocus

Commit:
Minor fixes.

Commit:
Made tests more modular.

Commit:
Added soft test.

Commit:
Corrected SRLS tests.

Commit:
Added first lateration fixes, phi_prime test to be done.

Commit:
Make acd input and result more consistent.

Commit:
Added ACD test.

Commit:
Added 'little noise' test

Commit:
Added new aloc file.

Commit:
Simplified angles reconstruction.

Commit:
Added new E-MDS variations.

Commit:
Added exceptions.

Commit:
Fixed AngleSet and extracted methods from procrustes.

Commit:
Removed wrongly failling assertion.

Commit:
Improved options parsing.

Commit:
Added unit tests.

Commit:
Updated E-MDS to test all variants.

Commit:
Merge pull request #5 from LCAV/algorithms/new_emds

Algorithms/new emds

Commit:
Improved readability of ACD and added unit tests.

Commit:
Added test for dwMDS, all tests passing.

Commit:
Added noisy test, made ACD faster.

Commit:
Added new norm function, improved plots.

Commit:
Customized data plots.

Commit:
Merge pull request #7 from LCAV/algorithms/iterative_methods

Algorithms/iterative methods

Commit:
Merge branch 'master' of github.com:LCAV/pylocus

Commit:
Fixed SRLS unit test.

Commit:
Fixed handling of nan values in SRLS.

Commit:
Reformatting.

Commit:
Fixing default solver option.

Commit:
Removed SCS solver.

fakufaku committed:
Added the option to rescaled the measurements in SRLS.

Commit:
Updated version.

fakufaku committed:
Updated description of rescale parameter in SRLS

Commit:
Fixed SRLS tests.

Commit:
Merge pull request #12 from fakufaku/rescaled_srls

Added the option to rescale the measurements in SRLS.

Commit:
Made ACD test work again.

Commit:
Relaxed numerical precision for test.

Commit:
Removed line add.

Commit:
Merge pull request #9 from LCAV/algorithms/improve_acd

Add new ACD test

Commit:
Updated branch based on master.

Commit:
Incorporated rescaling function in code.

Commit:
Merge branch 'master' into algorithms/srls_fix_nans

Commit:
Merge pull request #8 from LCAV/algorithms/srls_fix_nans

SRLS with rescale

Commit:
Added explanation.

Commit:
Merge pull request #13 from LCAV/fix/solver_choice

Fix/solver choice

Commit:
Added option to fix z in SRLS.

Commit:
Added iterative completion methods.

Commit:
Merge pull request #15 from LCAV/features/srls_2D

Added option to fix z in SRLS.

Commit:
Added passing more tests.

Commit:
Merge pull request #16 from LCAV/features/iterative_completion

Create iterative completion methods

Commit:
Added new options to edm completion.

Commit:
Fixed multisolution stepsize.

Commit:
Fixed ACD For 1-length step sizes.

Commit:
Fixed namings and added version number.

Commit:
Fixed naming.

Commit:
Reformatting and documentation.

Commit:
Updated documentation.

Commit:
Removed unused TODOs.

Commit:
Merge pull request #17 from LCAV/algorithms/improve_cdm

Algorithms/improve cdm

Commit:
Updated release number.

Commit:
Fixed version variable.

Commit:
Added cvxpy to requirements.

Commit:
Changed back to 0.0.1 version. (unstable)

oleksandr-pavlyk committed:
ENH: Efficiency improvements to routines

1. Achived 5X speed improvement in `mse` by using `np.square`
   rather than `np.power(x, 2)`, and by reusing intermediate buffers
   via use of universal function's out= keyword.

```
In [2]: # %load mse.py
   ...: import numpy as np
   ...:
   ...: def mse(x, xhat):
   ...:     """ Calcualte mse between vector or matrix x and xhat """
   ...:     sum_ = np.sum(np.power(x - xhat, 2))
   ...:     return sum_ / x.size
   ...:
   ...: def mse2(x, xhat):
   ...:     """ Calculate mse between vector or matrix x and xhar """
   ...:     sum_ = np.sum(np.square(x - xhat))
   ...:     return sum_ / x.size
   ...:
   ...: def mse3(x, xhat):
   ...:     """ Calcualte mse between vector or matrix x and xhat """
   ...:     sum_ = x - xhat
   ...:     np.square(sum_, out=sum_)
   ...:     return np.sum(sum_) / x.size
   ...:
   ...:

In [3]: v = np.random.uniform(0,1, size=10**6)

In [4]: m = np.random.uniform(0,1, size=(2048, 2048))

In [5]: %timeit mse(v, 0.5)
14.7 ms ± 6.88 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [6]: %timeit mse2(v, 0.5)
4.4 ms ± 2.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [7]: %timeit mse3(v, 0.5)
1.58 ms ± 734 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [8]: %timeit mse(m, 0.5)
67 ms ± 44.6 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [9]: %timeit mse2(m, 0.5)
24.7 ms ± 37.5 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [10]: %timeit mse3(m, 0.5)
15.1 ms ± 113 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [11]: np.equal(mse(v, 0.5), mse2(v, 0.5))
Out[11]: True

In [12]: np.equal(mse(v, 0.5), mse3(v, 0.5))
Out[12]: True

In [13]: np.all(np.equal(mse(m, 0.5), mse3(m, 0.5)))
Out[13]: True

In [14]: np.all(np.equal(mse(m, 0.5), mse2(m, 0.5)))
Out[14]: True

In [15]: def mse4(x, xhat):
    ...:     """ Calcualte mse between vector or matrix x and xhat """
    ...:     buf_ = x - xhat
    ...:     np.square(buf_, out=buf_)
    ...:     sum_ = np.sum(buf_)
    ...:     sum_ /= x.size
    ...:     return sum_
    ...:
    ...:

In [16]: %timeit mse4(m, 0.5)
14.4 ms ± 9.14 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [17]: %timeit mse4(v, 0.5)
1.6 ms ± 392 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [18]: %timeit mse4(v, 0.5)
1.59 ms ± 430 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

2. Routine `get_rotation_matrix` was computing composition of 3 rotations
   in the YZ plane through dot product of 3x3 matrices.

   Since it is equivalent to rotation in YZ plane though the sum of 3
   angles, construct the rotation matrix directly.

   If the routine was intended to construct rotation matrix by Euler angles,
   it did not do it right.

   I left the old code commented out.

3. Modified other routines to utilize intermediate buffers to speed them up.

oleksandr-pavlyk committed:
ENH: Replaced np.power(x,2) with np.square

oleksandr-pavlyk committed:
ENH: More efficients MDS function

Less copying, use of broadcast_to to avoid creation of intermediate
arrays.

Improved efficiency of `theta_from_eigendecomp` and `x_from_eigendecom`
by means of efficient use of broadcasting in numpy

oleksandr-pavlyk committed:
removed needless memory copy

oleksandr-pavlyk committed:
Removed comment per PR feedback

Avoided creation of unneeded copy of `x_est` in mds.py by
performing subtraction of its minimum in place.

Commit:
Merge pull request #18 from oleksandr-pavlyk/speed-impr

Speed improvements

Commit:
changed norm to L1

Commit:
remove .idea/ from bacjup

Commit:
remove .idea/ from bacjup

Commit:
add fixed_height z in execute_method

Commit:
added case n<d if z is fixed

Commit:
Added relative paths to enable submodules. Created version 0.0.2.

Commit:
Merge branch 'master' of github.com:LCAV/pylocus

Commit:
Made adjustments for new cvxpy version and for pytest.

Commit:
Made adjustments for new cvxpy version and for pytest.

Commit:
Made tests pass with pytest, changed travis setup.

Commit:
Added matplotlib requirement.

Commit:
Fixed cvxpy for this release.

Commit:
Made E-MDS run faster.

Commit:
Added build tag.

Commit:
Relaxed epsilon slightly.

Commit:
Merge branch 'fix/update_cvxpy_version' of https://github.com/LCAV/pylocus into fix/update_cvxpy_version

Commit:
Improved performance of SRLS.

Commit:
Merge pull request #20 from LCAV/fix/update_cvxpy_version

Made adjustments for new cvxpy version.

Commit:
Merge pull request #21 from LCAV/fix/srls_performance

Improved performance of SRLS.

Commit:
Upgraded to version 0.0.3.

Commit:
Renamed real_points to all_points.

Commit:
Renaming noisy_edm.

Commit:
Fixed undefined.

Commit:
Added more documentation.

Commit:
merge upstream

Ceow committed:
Delete modules.xml

Ceow committed:
Delete misc.xml

Ceow committed:
Delete pylocus.iml

Ceow committed:
Delete workspace.xml

Commit:
Added conda instructions.

Commit:
Added new tutorials folder and extra weighting matrix function.

Commit:
Add explanation in README file.

Commit:
Merge pull request #22 from LCAV/features/naming_and_doc

Features/naming and doc

Commit:
Merge branch 'master' of github.com:lcav/pylocus

Commit:
removed .orig files.

Commit:
Fixed indentation.

Commit:
Merge pull request #23 from Ceow/master

Little fixes

Commit:
Fixed required number for fixed height.

Commit:
Removed n parameter.

Commit:
Little fixes.

Commit:
Added optional requirements file, made cvxpy requirement 'soft' .

Commit:
Updated changelog.

Commit:
Removed noisy test, fixed dependencies in travis.

Commit:
Fixed newton.

Commit:
Improved import.

Commit:
Split SRLS up in different versions for readibility.

Commit:
Little extra assert.

Commit:
Added some comments.

Commit:
Introduced new Exception.

Commit:
Created new release.

Commit:
Remove duplicate version info.

Commit:
Merge pull request #24 from LCAV/feature/srls_improvement

Algorithm improvements

Commit:
Merge branch 'master' into algorithms/aloc

Commit:
Improved and added angle functionalities.

Commit:
Merge branch 'master' into angles_icassp

Commit:
Merge branch 'master' into algorithms/aloc

Commit:
Deleted aloc.py file because it is not useful anymore.

Commit:
Merge branch 'algorithms/aloc' into angles_icassp

Commit:
Fixed cvxpy requirement.

Commit:
Removed unused angle functionalities.

Commit:
Fix project description.

Commit:
Merge pull request #25 from LCAV/angles_icassp

Fix angle functionalities

Commit:
Fix the docs.

Commit:
Added basics to docs and removed unused modules.

Commit:
Add documentation for more basic functions.

Commit:
Add CHANGELOG entry and docs status to README.

Commit:
Merge pull request #26 from LCAV/doc_fix

Fix the docs

Commit:
Attempt to fix sphinx.

Commit:
Merge branch 'master' of github.com:LCAV/pylocus

Commit:
Add .readthedocs.yml.

Commit:
Change printout
